### PR TITLE
add serialport

### PR DESCRIPTION
**Pull request by JerryYangKai**

### CODE REVIEW

File: src/Models/AZ3166Device.ts [hunk 1]
```diff
@@ -14,7 +14,7 @@ import * as WinReg from 'winreg';
 
 import {BoardProvider} from '../boardProvider';
 import {ConfigHandler} from '../configHandler';
-import {ConfigKey} from '../constants';
+import {ConfigKey, ScaffoldType} from '../constants';
```
Comment (dilin-MS) [inline src/Models/AZ3166Device.ts]: >ScaffoldType [](start = 19, length = 12)

ScaffoldType seems not needed in this file.

**Comment by dilin-MS** [comment]
>     ];

Move this code to templated files under resource/templates for consolidation.

---
Refers to: src/Models/AZ3166Device.ts:175 in 41b1f8d. [](commit_id = 41b1f8dccf44787b82fae9148a0b5f41b3beb265, deletion_comment = False)

**Comment by dilin-MS** [comment]
>       return false;

What's the return value meaning of this fuction? What are true/false/undefined represent? Try to handle "user cancel operation" using the CancelOperationError.

---
Refers to: src/Models/AZ3166Device.ts:186 in 41b1f8d. [](commit_id = 41b1f8dccf44787b82fae9148a0b5f41b3beb265, deletion_comment = False)

**Comment by dilin-MS** [comment]
>         ];

Try to put this part of code into a sub function to reduce the main function lines.

---
Refers to: src/Models/AZ3166Device.ts:227 in 41b1f8d. [](commit_id = 41b1f8dccf44787b82fae9148a0b5f41b3beb265, deletion_comment = False)

**Comment by dilin-MS** [comment]
>   async configDeviceSettings(): Promise<boolean> {

This function is too long. Please try to break it down for better human readability.

---
Refers to: src/Models/AZ3166Device.ts:151 in 41b1f8d. [](commit_id = 41b1f8dccf44787b82fae9148a0b5f41b3beb265, deletion_comment = False)

**Comment by JerryYangKai** [comment]
comment   'return false;' 
will refactor next pr

File: src/Models/AZ3166Device.ts [hunk 2]
```diff
@@ -148,24 +149,23 @@ export class AZ3166Device extends ArduinoDeviceBase {
...
-        detail: 'Config CRC'
-      }
-    ];
+    // Read options configuration JSON
+    const devciceConfigFilePath: string =
+        this.extensionContext.asAbsolutePath(path.join(
+            FileNames.resourcesFolderName, FileNames.templatesFolderName,
+            FileNames.configDeviceOptionsFileName));
+    const configSelectionItemsContent =
+        JSON.parse(fs.readFileSync(devciceConfigFilePath, 'utf8'));
```
Comment (dilin-MS) [inline src/Models/AZ3166Device.ts]: >fs.readFileSync( [](start = 19, length = 16)

Use FileUtility.readFile for consolidation.

File: src/Models/AZ3166Device.ts [hunk 3]
```diff
@@ -188,46 +188,8 @@ export class AZ3166Device extends ArduinoDeviceBase {
...
-        deviceConnectionStringSelection = [{
-          label: 'Input IoT Hub Device Connection String',
-          description: '',
-          detail: ''
-        }];
-      }
-
       const selection = await vscode.window.showQuickPick(
-          deviceConnectionStringSelection,
+          this.getDeviceConnectionStringOptions(deviceConnectionString),
```
Comment (dilin-MS) [inline src/Models/AZ3166Device.ts]: >this.getDeviceConnectionStringOptions(deviceConnectionString) [](start = 10, length = 61)

Use a variable 'deviceConnectionStringSelection' to store the result of this function and pass the variable as parameter to showQuickPick() function for readability.

**Comment by dilin-MS** [comment]
>         });

This part can be broke down to a sub function such as getConfigSelection().

---
Refers to: src/Models/AZ3166Device.ts:176 in 2454ed0. [](commit_id = 2454ed08a6774c4eba980e0d8682f41d97083568, deletion_comment = False)

**Comment by dilin-MS** [comment]
>       return false;

Throw "userCancelOperationError" if user cancel the selection.

---
Refers to: src/Models/AZ3166Device.ts:179 in 2454ed0. [](commit_id = 2454ed08a6774c4eba980e0d8682f41d97083568, deletion_comment = False)

File: src/Models/AZ3166Device.ts (hunk 3 above)
Comment (dilin-MS) [inline src/Models/AZ3166Device.ts]: What if user cancel the device connection string selection? Handle the cancel situation using the "userCancelOperationError".

**Comment by dilin-MS** [comment]
>         return false;

Same here.

---
Refers to: src/Models/AZ3166Device.ts:196 in 2454ed0. [](commit_id = 2454ed08a6774c4eba980e0d8682f41d97083568, deletion_comment = False)

**Comment by dilin-MS** [comment]
>       }

Each selection can be moved to small function so the main function does not look so long. And the purpose and step of the function can be more clear.

---
Refers to: src/Models/AZ3166Device.ts:231 in 2454ed0. [](commit_id = 2454ed08a6774c4eba980e0d8682f41d97083568, deletion_comment = False)

**Comment by dilin-MS** [comment]
>         return false;

Same here.

---
Refers to: src/Models/AZ3166Device.ts:234 in 2454ed0. [](commit_id = 2454ed08a6774c4eba980e0d8682f41d97083568, deletion_comment = False)

File: src/Models/AZ3166Device.ts [hunk 4]
```diff
@@ -355,6 +328,65 @@ export class AZ3166Device extends ArduinoDeviceBase {
...
+    // Try to close serial monitor
+    try {
+      await vscode.commands.executeCommand(
+          'arduino.closeSerialMonitor', null, false);
+    } catch (ignore) {
+    }
+
+    // Set selected connection string to device
+    const plat = os.platform();
+    if (plat === 'win32') {
```
Comment (dilin-MS) [inline src/Models/AZ3166Device.ts]: >win32 [](start = 18, length = 5)

There is variable OSPlatform.WIN32 representing this constant.

File: src/Models/AZ3166Device.ts [hunk 5]
```diff
@@ -355,6 +328,65 @@ export class AZ3166Device extends ArduinoDeviceBase {
...
+      return await this.flushDeviceConfig(configValue, option);
+    } else {
```
Comment (dilin-MS) [inline src/Models/AZ3166Device.ts]: Does this situation apply to both macOS and unix?

**Comment by JerryYangKai** [comment]
@dilin-MS  I have committed 'code refactor', can you review it? Thanks a lot

File: src/Models/AZ3166Device.ts [hunk 6]
```diff
@@ -148,33 +150,14 @@ export class AZ3166Device extends ArduinoDeviceBase {
...
-      },
-      {
-        label: 'Generate CRC for OTA',
-        description:
-            'Generate Cyclic Redundancy Check(CRC) code for OTA Update',
-        detail: 'Config CRC'
-      }
-    ];
-
+    const configSelectionItems = this.getMainOptions();
```
Comment (dilin-MS) [inline src/Models/AZ3166Device.ts]: >getMainOptions [](start = 38, length = 14)

getConfigDeviceSettingsMainOptions()

**Pull request merged**